game/view/actor/attr.lua

```diff
@@ -0,0 +1,148 @@
+
+local RES     = require 'resources'
+local RANDOM  = require 'common.random'
+local APT     = require 'domain.definitions.aptitude'
+local COLORS  = require 'domain.definitions.colors'
+local FONT    = require 'view.helpers.font'
+local PLAYSFX = require 'helpers.playsfx'
+
+local abs  = math.abs
+local min  = math.min
+local max  = math.max
+local pi   = math.pi
+local fmod = math.fmod
+
+local coresume = coroutine.resume
+local cocreate = coroutine.create
+local coyield  = coroutine.yield
+
+local delta = love.timer.getDelta
+
+local _dt
+local _barwidth
+local _font
+local _rot = 0
+local _particles = {}
+local _states = {}
+
+local function _newParticleSource()
+  local pixel = RES.loadTexture('pixel')
+  local particles = love.graphics.newParticleSystem(pixel, 128)
+  particles:setParticleLifetime(.75)
+  particles:setSizeVariation(0)
+  particles:setLinearDamping(6)
+  particles:setSpeed(256)
+  particles:setSpread(2*pi)
+  particles:setColors(COLORS.NEUTRAL, COLORS.TRANSP)
+  particles:setSizes(4)
+  particles:setEmissionArea('ellipse', 0, 0, 0, false)
+  particles:setTangentialAcceleration(-512)
+  return particles
+end
+
+local function _renderAttribute(data)
+  local g = love.graphics
+  local actor, attrname, particles = unpack(data)
+  local progress = 0
+  local lvl = actor:getAttrLevel(attrname)
+  while true do
+    -- update bar progress
+    local upgrade = actor:getAttrUpgrade(attrname)
+    local aptitude = actor:getAptitude(attrname)
+    local total_prev = APT.CUMULATIVE_REQUIRED_ATTR_UPGRADE(aptitude, lvl)
+    local total_next = APT.CUMULATIVE_REQUIRED_ATTR_UPGRADE(aptitude, lvl+1)
+    local target = min(1 ,
+      max(0, (upgrade - total_prev) / (total_next - total_prev))
+    )
+    progress = min(
+      1, max(0, progress + 8*(target-progress)*_dt)
+    )
+    if abs(target - progress) < 0.01 then
+      progress = target
+    end
+
+    -- check if level up
+    local newlvl = actor:getAttrLevel(attrname)
+    local offset = newlvl - lvl
+    local step = offset / abs(offset)
+    if newlvl ~= lvl and progress == 1 then
+      lvl = lvl + step
+      offset = offset - step
+      progress = 0
+      particles:emit(48)
+      PLAYSFX('get-item')
+      local start = 0
+      local rand = RANDOM.safeGenerate
+      while start <= 0.5 do
+        g.translate(2*(rand()*2-1), 2*(rand()*2-1))
+        start = start + _dt
+        coyield(lvl, offset, progress, true)
+      end
+    end
+
+    -- yield
+    coyield(lvl, offset, progress)
+  end
+end
+
+local ATTR = {}
+
+function ATTR.init(width)
+  _barwidth = width/4
+  _font = FONT.get("Text", 20)
```

Comment:
These two are essentially constant (`_barwidth` maybe not so much, but even then). so could we write them in uppercase?

Reply:
there is nowhere else in the code where we use fonts as a constant. they are objects loaded and cached in runtime

Reply:
they can change in runtime. i think i mess with the lineheight in sometimes. so i don't think uppercasing them at all is a good idea

Reply:
All right, makes sense.